domains/client/relayer/src/lib.rs

```diff
@@ -0,0 +1,130 @@
+    /// Emits when failed to submit an unsigned extrinsic.
+    SubmitUnsignedExtrinsic,
+}
+
+impl<Client, Block> Relayer<Client, Block>
+where
+    Block: BlockT,
+    Client: HeaderBackend<Block>
+        + StateBackend<<Block::Header as HeaderT>::Hashing>
+        + ProofProvider<Block>
+        + ProvideRuntimeApi<Block>,
+    Client::Api: RelayerApi<Block, RelayerId, DomainId>,
+{
+    /// Constructs the proof for the given key using the backend for the given key.
+    fn construct_storage_proof_for_key_at(
+        &self,
+        block_id: &BlockId<Block>,
+        key: &StorageKey,
+    ) -> Result<Proof<Block::Hash>, Error> {
+        let state_version = sp_runtime::StateVersion::default();
+        let state_root = self
+            .domain_client
+            .storage_root(std::iter::empty(), state_version)
+            .0;
+        let proof = self
+            .domain_client
+            .read_proof(block_id, &mut [key.as_ref()].into_iter())
+            .map_err(|_| Error::ConstructStorageProof)?;
+
+        Ok(Proof {
+            state_root,
+            message_proof: proof,
+        })
+    }
+
+    fn construct_cross_domain_message_and_submit<
+        Submitter: Fn(CrossDomainMessage<DomainId, Block::Hash>) -> Result<(), sp_api::ApiError>,
+    >(
+        &self,
+        k_deep_block_id: &BlockId<Block>,
+        msgs: Vec<RelayerMessageWithStorageKey<DomainId>>,
+        submitter: Submitter,
+    ) -> Result<(), Error> {
+        for msg in msgs {
+            let proof =
+                match self.construct_storage_proof_for_key_at(k_deep_block_id, &msg.storage_key) {
+                    Ok(proof) => proof,
+                    Err(_) => {
+                        tracing::error!(
+                        target: LOG_TARGET,
+                        "Failed to construct storage proof for message: {:?} bound to domain: {:?}",
+                        (msg.channel_id, msg.nonce),
+                        msg.dst_domain_id,
+                    );
+                        continue;
+                    }
+                };
+            let msg = CrossDomainMessage::from_relayer_msg_with_proof(msg, proof);
+            let (dst_domain, msg_id) = (msg.dst_domain_id, (msg.channel_id, msg.nonce));
+            match submitter(msg).map_err(|_| Error::SubmitUnsignedExtrinsic) {
+                Ok(_) => {}
+                Err(_) => {
+                    tracing::error!(
+                        target: LOG_TARGET,
+                        "Failed to submit message: {:?} to domain: {:?}",
+                        msg_id,
+                        dst_domain,
+                    );
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn submit_unsigned_messages(&self, block_id: &BlockId<Block>) -> Result<(), Error> {
+        let best_block_id = BlockId::Hash(self.domain_client.info().best_hash);
+        let api = self.domain_client.runtime_api();
+        // get assigned messages
```

Comment:
The comments inside this function don't convey more information than the code itself.